### PR TITLE
Fixes #8755: Document node properties

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/8755

### CODE REVIEW

File: 4_advanced_usage/40_node_properties.txt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+- in the technique editor
+- in your own CFEngine code
+
+To use a property, simply use the variable node.properties with the variable call syntax.
+
+Example with a string property named mykey:
+
+----
+
+${node.properties[mykey]}
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Please use real-life examples wherever possible. "datacenter=Paris" is a good example (here, just replace "mykey" by "datacenter" twice)

File: 4_advanced_usage/40_node_properties.txt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+----
+
+
+==== Under the hood (CFEngine)
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Can you replace "CFEngine" with "policy generation"?

File: 4_advanced_usage/40_node_properties.txt [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+On the server, one or more properties files are written for each node in the /var/rudder/share/<uuid>/rules/cfengine-community/properties.d/ directory.
+This directory is then synchronized by the agent with all other promise files.
+
+In CFEngine, properties are made available in the node.namespace container that contains the values.
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Can you replace "CFEngine" with "the agent"?

Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Can you replace "node.namespace" with "node.<namespace>" to illustrate that this is placeholder name that will change?

File: 4_advanced_usage/40_node_properties.txt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+Those values are read from /var/rudder/cfengine-community/properties/*.json
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: I think this should be /var/rudder/cfengine-community/**inputs/**properties/*.json

File: 4_advanced_usage/40_node_properties.txt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+All files are taken in order and overrides the previous one, the last one wins.
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: s/overrides the previous one/override the previous ones/

File: 4_advanced_usage/40_node_properties.txt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Each file must contain at least a 2 levels JSON content, the first level is the namespace level
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: "Each file must contain JSON content with at least two levels, ..."

An example here would be useful.

File: 4_advanced_usage/40_node_properties.txt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+and the second level is the key level.
+The namespace name must comply with CFEngine variable names rules (ie [a-zA-Z_]+)
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Can you rephrase this to be "The namespace name must match the following regex: [a-zA-Z0-9_]+"

File: 4_advanced_usage/40_node_properties.txt [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+The merge is a first level merge done at the namespace level.
+This means that:
+- a key in a namespace is fully overridden by the same key in the same namespace in a later file.
+- a key in a namespace is never overriden by the same key in a different namespace
+- a key that is overriden never retain original data even if it is a data container itself
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: s/retain/retains/

File: 4_advanced_usage/40_node_properties.txt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+The result key is available in the node.<namespace> data variable
+usage example: ${node.properties[my_key]}
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Again, here, please use a real-world example such as ```${node.properties[datacenter]}``` (or customer, etc)

File: 4_advanced_usage/40_node_properties.txt [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+To get the original data (for debug only) there is the properties.property_<fileid> variable.
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: s/there is/you can use/

File: 4_advanced_usage/40_node_properties.txt [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+Node properties can be modified using Rudder API, see http://www.rudder-project.org/rudder-api-doc/#api-Nodes-updateNodeProperties
+
+Currently a property is interpreted as a string value. So properties can be seen as a list of key=value pairs.
+
+==== Using properties
+
+You can use a node property almost everywhere in Rudder:
+- in directive parameters
+- in the technique editor
+- in your own CFEngine code
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Can you replace "CFEngine code" by "techniques and generic methods"?

**Comment by peckpeck** [comment]
Commit modified

File: 4_advanced_usage/40_node_properties.txt [hunk 12]
```diff
@@ -0,0 +1,65 @@
+=== Node properties
+
+==== Defining properties
+
+Node properties can be found in the "properties" tab of each node in Rudder.
+
+Node properties can be modified using Rudder API, see http://www.rudder-project.org/rudder-api-doc/#api-Nodes-updateNodeProperties
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Does this appear as a link when the docs are built?

Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: If not, please make it a link.

File: 4_advanced_usage/40_node_properties.txt [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+To use a property, simply use the variable node.properties with the variable call syntax.
+
+Example with a string property named datacenter:
+
+----
+
+${node.properties[datacenter]}
+
+----
+
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: We should add a warning here:

Currently node properties can not be used inside ${eval } JavaScript blocks, since they are evaluated during poliocy generation and node properties are only made available in the agents. A workaround for this will be provided in a future version of Rudder.

File: 4_advanced_usage/40_node_properties.txt [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+==== Under the hood
+
+On the server, one or more properties files are written for each node in the /var/rudder/share/<uuid>/rules/cfengine-community/properties.d/ directory.
+This directory is then synchronized by the agent with all other promise files.
+
+In the agent, properties are made available in the node.<namespace> container that contains the values.
+Those values are read from /var/rudder/cfengine-community/inputs/properties/*.json
+All files are taken in order and override the previous ones, the last one wins.
+
+Each file must contain at least a 2 levels JSON content, the first level is the namespace level
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: s/a 2 levels JSON content/2 levels of JSON content/

File: 4_advanced_usage/40_node_properties.txt [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+and the second level is the key level.
+The namespace name must be an ASCII name that doesn't start with '_', (it must match the following regex: [a-zA-Z0-9][a-zA-Z0-9_]*).
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: ", (" is an odd use of punctuation :)

How about "The namespace name must be an ASCII name that doesn't start with '_' and must match the following regex: [a-zA-Z0-9][a-zA-Z0-9_]*." ?

File: 4_advanced_usage/40_node_properties.txt [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+Example: 
+
+----
+
+{ 
+  "properties": 
+  {
+    "datacenter": "Paris"
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Could we add a couple more key-value pairs in the example?

"datacenter": "Paris",
"environment": "production",
"customer": "Normation"

File: 4_advanced_usage/40_node_properties.txt (hunk 12 above)
Comment (peckpeck) [inline 4_advanced_usage/40_node_properties.txt]: I don't know

Comment (peckpeck) [inline 4_advanced_usage/40_node_properties.txt]: I cannot generate the doc at the moment, but it seems to work for other links in the documentation.

**Comment by peckpeck** [comment]
Commit modified

File: 4_advanced_usage/40_node_properties.txt [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+Example with a string property named datacenter:
+
+----
+
+${node.properties[datacenter]}
+
+----
+
+WARNING: currently node properties can not be used inside ${eval } JavaScript blocks, since they are evaluated during poliocy generation and node properties are only made available in the agents. A workaround for this will be provided in a future version of Rudder.
```
Comment (jooooooon) [inline 4_advanced_usage/40_node_properties.txt]: Oops, I made a typo, sorry: s/poliocy/policy/

**Comment by jooooooon** [comment]
This looks good to me apart from that one typo. Please fix and self-merge.

**Comment by peckpeck** [comment]
Commit modified
